src/main/host/syscall/unistd.rs

```diff
@@ -27,32 +29,42 @@ pub fn close(sys: &mut c::SysCallHandler, args: &c::SysCallArgs) -> c::SysCallRe
         };
 
         // if it's a legacy descriptor, use the C syscall handler instead
-        if let CompatDescriptor::Legacy(_) = desc {
-            return unsafe {
-                c::syscallhandler_close(
+        let desc = match desc {
+            CompatDescriptor::New(d) => d,
+            CompatDescriptor::Legacy(_) => unsafe {
+                return c::syscallhandler_close(
                     sys as *mut c::SysCallHandler,
                     args as *const c::SysCallArgs,
-                )
-            };
+                );
+            },
+        };
+
+        // if this is the last remaining descriptor
+        if desc.get_fd_count() == 1 {
```

Comment:
Could this be implemented by implementing Drop, instead of keeping an internal explicit count?

Reply:
I had started doing it this way, but the problem is that you can't pass arguments to `Drop::drop()`. My two concerns are:

1. Since calling `close()` on a descriptor/file object may need to emit events (for example if it sets its status flag to `CLOSED`), it needs access to an `EventLoop` object to add the event to. We could make a new `EventLoop` object within the descriptor's `Drop::drop()` implementation and process any events there immediately, but this drop could happen while we're already holding references to other objects, in which case we could end up trying to borrow those same objects within the `Drop::drop()` and end up with a runtime error.
2. Right now we only pass an `EventLoop` struct to `Descriptor::close()`, but if we need to pass the Host object or Process object in the future, then that will be a problem.

I'd be okay with using the `Drop::drop()` for now if you think it's better, as the above are only hypothetical issues at this point.

Reply:
>it needs access to an EventLoop object to add the event to. We could make a new EventLoop object within the descriptor's Drop::drop() implementation and process any events there immediately,

Seems fine, particularly since that's what we're doing here as well...?

> this drop could happen while we're already holding references to other objects, in which case we could end up trying to borrow those same objects within the Drop::drop() and end up with a runtime error.

I think this possibility isn't too bad. Unlike, say, Java, in Rust objects are dropped at a deterministic time (roughly when they go out of scope). So I *think* it'd end up being close to equivalent to explicitly doing the cleanup here, or at least if it failed it'd fail predictably and could reasonably be debugged.

> Right now we only pass an EventLoop struct to Descriptor::close(), but if we need to pass the Host object or Process object in the future, then that will be a problem.

`drop` also gets a reference to `self`. Should the `Host` and `Process` be reachable from there?

All that said, from our conversation on slack, if we're keeping an explicit count here instead of creating a separate object, it seems like a Drop won't work.

I'm ok with this ending up being a special case where it's easier to keep an explicit count and do explicit cleanup; I just don't want it to become the default :)